tests/FileSystemTests.cpp

```diff
@@ -773,6 +773,64 @@ TEST_CASE("FileSystemTests::CreateFileW helpers", "[filesystem]")
 }
 
 #endif
+TEST_CASE("FileSystemTest::FolderChangeReader destructor does not hang", "[filesystem]")
+{
+    std::wstring testRootDir;
+    testRootDir = wil::ExpandEnvironmentStringsW<std::wstring>(L"%TEMP%\\wil_test_filesystem");
+    std::wstring testFile = testRootDir + L"\\test.dat";
+    bool deleteDir = false;
+    wil::unique_handle opCompleteEv(::CreateEventW(nullptr, TRUE, FALSE, nullptr));
+    wil::unique_handle wilReaderNotify(::CreateEventW(nullptr, TRUE, FALSE, nullptr));
+
+    REQUIRE_FALSE(DirectoryExists(testRootDir.c_str()));
+    REQUIRE(SUCCEEDED(wil::CreateDirectoryDeepNoThrow(testRootDir.c_str())));
+    REQUIRE(DirectoryExists(testRootDir.c_str()));
+
+    /**
+     * Move to a new thread.
+     * The destructor of unique_folder_change_reader might hang. If this happens, 
+     * we want to report an test error instead of hanging forever. 
+     */
+    auto reader = wil::make_folder_change_reader_nothrow(testRootDir.c_str(), false, wil::FolderChangeEvents::All,
+    [&](wil::FolderChangeEvent, PCWSTR)
+    {
+        if (deleteDir)
+            RemoveDirectoryW(testRootDir.c_str());
 
+        SetEvent(opCompleteEv.get());
+    });
+
+    struct ThreadContext

```

Comment:
I cannot use `std::thread` because `join()` doesn't have any timeout. However, I will rewrite it using `std::future` / `std::async` to be able to `wait_for(...)` the result

Reply:
actually i cannot use `std::future` because I cannot detach it in case of a hang. So I rewrote the code using `std::thread` + `native_handle()`